block-storage/src/main/scala/coop/rchain/blockstorage/dag/InMemBlockDagStorage.scala

```diff
@@ -46,8 +46,16 @@ final class InMemBlockDagStorage[F[_]: Concurrent: Sync: Log: BlockStore](
       dataLookup.contains(blockHash).pure[F]
     def lookupByDeployId(deployId: DeployId): F[Option[BlockHash]] =
       blockHashesByDeploy.get(deployId).pure[F]
-    def topoSort(startBlockNumber: Long): F[Vector[Vector[BlockHash]]] =
-      topoSortVector.drop(startBlockNumber.toInt).pure[F]
+    def topoSort(
+        startBlockNumber: Long,
+        maybeEndBlockNumber: Option[Long]
+    ): F[Vector[Vector[BlockHash]]] = {
+      val endBlockNumber: Long = maybeEndBlockNumber.getOrElse(topoSortVector.length.toLong)
+      topoSortVector
+        .take(endBlockNumber.toInt)
+        .takeRight((endBlockNumber - startBlockNumber + 1).toInt)
+        .pure[F]
+    }
     def topoSortTail(tailLength: Int): F[Vector[Vector[BlockHash]]] =
       topoSortVector.takeRight(tailLength).pure[F]
```

Comment:
I don't understand how to do that on the `extension function` you referred to in codes aspect.
Can you describe more?

Because we got two implementations for `BlockDagRepresentation`, the `FileBlockDagRepresentation` is not like that.

Reply:
`topoSortTail` can be written in terms of `topoSort(0, depth)` so it doesn't have to be directly on the interface.

Reply:
Sorry, I'm wrong about this. For `topoSortTail` we need the length of _topoSortVector_ so it cannot be just rewritten with `topoSort`.